### PR TITLE
Implement some gz functions

### PR DESCRIPTION
**Pull request by rzinsly**
This PR adds NX versions for gzopen, gzclose, gzwrite and gzread.
I'm hitting CC 66 with `nx_gzread` both on libnxz and zlib compressed files:
```
pid 2855471: Info: uncompress begin: len 16384
pid 2855471: nx_inflate.c:1367: Error: error: cc = 66, cc = 0x42
pid 2855471: nx_inflate.c:1370: Error: CSB: 0x 80 00 42 60
pid 2855471: nx_inflate.c:1370: Error: CSB: 0x 80 00 42 60
pid 2855471: nx_inflate.c:1370: Error: CSB: 0x 80 00 42 60
pid 2855471: nx_inflate.c:1370: Error: CSB: 0x 80 00 42 60
pid 2855471: nx_inflate.c:1690: Error: rc -3
```
I'll also add tests for these new functions.

### CODE REVIEW

File: lib/nx_gzlib.c [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+		state->fd = fopen(path, mode);
+	if(state->fd == NULL){
+		free(state);
+		return NULL;
+	}
+
+	strm = &(state->strm);
+	memset(strm, 0, sizeof(z_stream));
+	if(strchr(mode, 'w')){
+		err = nx_deflateInit2_(&(state->strm), Z_DEFAULT_COMPRESSION, Z_DEFLATED, 31, DEF_MEM_LEVEL, Z_DEFAULT_STRATEGY, ZLIB_VERSION, (int)sizeof(z_stream));
```
Comment (tuliom) [inline lib/nx_gzlib.c:94]: This is a very long line.
Could you limit it to 80 characters, please?

File: lib/nx_gzlib.c [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp strm = &(state->strm);
+	int ret;
+
+	if(file == NULL) return Z_STREAM_ERROR;
+
+	ret = nx_inflateEnd(strm);
+
+	return ret;
```
Comment (tuliom) [inline lib/nx_gzlib.c:60]: I expected to see a `free(file)` somewhere here.

File: lib/nx_gzlib.c [hunk 3]
```diff
@@ -103,3 +103,24 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
...
 	return file;
 }
+
+int nx_gzwrite (gzFile file, const void *buf, unsigned len)
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	int rc;
+	uLong last_total_in = stream->total_in;
+	unsigned char* next_out = malloc(sizeof(char)*len);
```
Comment (tuliom) [inline lib/nx_gzlib.c:136]: Where is this being freed?

File: lib/nx_gzlib.c [hunk 4]
```diff
@@ -104,6 +104,29 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
 	return file;
 }
 
+int nx_gzread(gzFile file, void *buf, unsigned len)
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	int ret, err;
+	uLong last_total_out = stream->total_out;
+	unsigned char* next_in = malloc(sizeof(char)*len);
```
Comment (tuliom) [inline lib/nx_gzlib.c:113]: Where is this being freed?

**Review by tuliom — CHANGES_REQUESTED** [review]

File: lib/nx_gzlib.c (hunk 2 above)
Comment (rzinsly) [reply lib/nx_gzlib.c:60]: Yes, I'll fix that for the next version.

File: lib/nx_gzlib.c (hunk 3 above)
Comment (rzinsly) [reply lib/nx_gzlib.c:136]: It isn't, I'll fix that for the next version.

File: lib/nx_gzlib.c (hunk 1 above)
Comment (mscastanho) [reply lib/nx_gzlib.c:94]: @rzinsly I added `nx_deflateInit2` as part of an upcoming PR, but I split the commit into #139 in case you want to use it here. It will help make this line shorter.

**Comment by rzinsly** [comment]
I have a functional version of the patches now, however there is two potential issues:

1. We don't support different strategies now, so if someone set them on `gzopen` `deflateInit` will fail, perhaps we should not set them yet.
2. We don't know how much source will be needed on `gzread` to decompress `len` bytes, reading more from the file will make `gzread` faster however if it's too much we will overflow `fifo_in`. I'm using 10 because it's safe as even sparse1000M doesn't overflow that and it is unlikely there will be a file that compress more than sparse1000M. Reading one byte per loop and inflating with `Z_SYNC_FLUSH` is the safest option but it is slow.

File: test/libnxz.abi [hunk 5]
```diff
@@ -24,6 +24,11 @@
     <elf-symbol name='deflateResetKeep' version='ZLIB_1.2.5.2' is-default-version='yes' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
     <elf-symbol name='deflateSetDictionary' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
     <elf-symbol name='deflateSetHeader' version='ZLIB_1.2.2' is-default-version='yes' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='gzclose' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='gzdopen' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='gzopen' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='gzread' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='gzwrite' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
```
Comment (tuliom) [inline test/libnxz.abi:31]: I'm afraid this patch has to be split and its content to be distributed across the previous patches in order to guarantee the previous patches do not break `git-bisect`.

**Review by tuliom — CHANGES_REQUESTED** [review]

File: lib/nx_gzlib.c [hunk 6]
```diff
@@ -155,6 +156,44 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
...
+		if (err != Z_OK && err != Z_STREAM_END){
+			state->err = err;
+			return 0;
+		}
+	} while (stream->avail_out != 0);
+
+	if (stream->avail_in > 0){
+		nx_streamp s = (nx_streamp) stream->state;
+		copy_data_to_fifo_in(s);
+	}
```
[the remark this answers is not in the document]
Comment (rzinsly) [reply lib/nx_gzlib.c:191]: Because `gzread` returns when `len` bytes are wrote on `buf`, if we still have unprocessed data on `avail_in` we need to save it or we will lose it, IMHO `fifo_in` seems the best place to save it. Another option is to skip `read` when `avail_in` is not 0, but we can't guarantee `avail_in` will not be overwritten somewhere else between `gzread` calls.

File: lib/nx_gzlib.c [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	int err;
+	z_stream strm;
+	bool deflate;
+} *gz_statep;
+
+int nx_gzclose(gzFile file)
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	unsigned char* next_out = malloc(sizeof(char)*MAX_LEN);
```
[the remark this answers is not in the document]
Comment (rzinsly) [reply lib/nx_gzlib.c:62]: Oops, fixed!

File: lib/nx_gzlib.c [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+		/* Finish the stream.  */
+		stream->next_in = Z_NULL;
+		stream->avail_in = 0;
+		do {
+			stream->next_out = next_out;
+			stream->avail_out = MAX_LEN;
+
+			nx_deflate(stream, Z_FINISH);
+			write(state->fd, next_out, MAX_LEN - stream->avail_out);
+		} while (stream->avail_out == 0);
```
Comment (mscastanho) [inline lib/nx_gzlib.c:80]: If the pending data is exactly `MAX_LEN`, this will execute an extra unneeded loop iteration.

It would be better to change this check to `while (ret != Z_STREAM_END)`, where `ret` would be the return value from `nx_deflate`.

File: lib/nx_inflate.c [hunk 9]
```diff
@@ -1692,13 +1658,18 @@ int nx_inflateSetDictionary(z_streamp strm, const Bytef *dictionary, uInt dictLe
 	if (NULL == (s = (nx_streamp) strm->state))
 		return Z_STREAM_ERROR;
 
+	if (s->wrap == (HEADER_ZLIB | HEADER_GZIP)) {
+		prt_err("inflateSetDictionary error: inflate did not dected the header yet.\n");
```
Comment (mscastanho) [inline lib/nx_inflate.c:1662]: I suspect by `dected` you meant `detect`?

File: lib/nx_zlib.h [hunk 10]
```diff
@@ -113,11 +113,12 @@ struct selector {
...
 #define HEADER_GZIP  2
 
+/* 47 is the max windowBits for auto detect header */
 #ifndef MAX_WBITS
-#  define MAX_WBITS 15
+#  define MAX_WBITS 47
 #endif
 #ifndef DEF_WBITS
-#  define DEF_WBITS MAX_WBITS
+#  define DEF_WBITS 47
```
Comment (mscastanho) [inline lib/nx_zlib.h:121]: Is this change necessary? The new value of `MAX_WBITS` would have set `47` here as well.

File: lib/nx_gzlib.c [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+gzFile nx_gzopen_(const char* path, int fd, const char *mode)
+{
+	gz_statep state;
+	gzFile file;
+	int err, strategy = Z_DEFAULT_STRATEGY, level = Z_DEFAULT_COMPRESSION;
+	char* digit;
+	void* stream;
+
+	state = malloc(sizeof(struct gz_state));
+	memset(state, 0, sizeof(struct gz_state));
```
Comment (mscastanho) [inline lib/nx_gzlib.c:111]: This line is being added in the 4th commit. Doesn't it belong with the first one that adds `nx_gzopen`?

File: lib/nx_gzlib.c [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+	file = (gzFile) state;
+	return file;
+}
+
+int nx_gzread(gzFile file, void *buf, unsigned len)
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	int ret, err;
+	const uInt max = len > 10? 10 : 1;
```
Comment (mscastanho) [inline lib/nx_gzlib.c:165]: Is this correct? Why are you using such low values?

File: lib/nx_gzlib.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mscastanho) [reply lib/nx_gzlib.c:192]: @rzinsly another option is to `lseek` the file offset `avail_in` bytes back, so the next `read` will get the unprocessed bytes. The downside is that we would be re-reading these bytes from disk in the future, while a copy in memory would most likely be cheaper.

**Review by mscastanho — CHANGES_REQUESTED** [review]
Haven't looked at the test yet, but some early comments below.

File: lib/nx_gzlib.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (tuliom) [reply lib/nx_gzlib.c:192]: I like what @rzinsly did here, but I think the code deserves a source code comment that is a copy if his explanation here.

File: lib/nx_gzlib.c [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+ *
+ */
+#include "nx_zlib.h"
+
+#define DEF_MEM_LEVEL 8
+
+gzFile nx_gzopen_(const char* path, int fd, const char *mode);
+
+typedef struct gz_state {
+	gzFile gz;
```
[the remark this answers is not in the document]
Comment (rzinsly) [reply lib/nx_gzlib.c:51]: I'm not sure I understand your question.
`gzFile` is a pointer to the semi-opaque structure `gzFile_s`, and it is available on `zlib.h`. What is not available is `gz_state`, that is the full file state data structure that we are defining here.

File: lib/nx_zlib.h (hunk 10 above)
Comment (rzinsly) [reply lib/nx_zlib.h:121]: Seems zlib is defining `MAX_WBITS` somewhere. Without this change, at least in the servers I tested, `DEF_WBITS` is always 15 and libnxz assumes we have a zlib header.

File: lib/nx_gzlib.c (hunk 12 above)
Comment (rzinsly) [reply lib/nx_gzlib.c:165]: Yes, that's one of the points I raised on https://github.com/libnxz/power-gzip/pull/138#issuecomment-1077786216

**Comment by tuliom** [comment]
>     2. We don't know how much source will be needed on `gzread` to decompress `len` bytes, reading more from the file will make `gzread` faster however if it's too much we will overflow `fifo_in`. I'm using 10 because it's safe as even sparse1000M doesn't overflow that and it is unlikely there will be a file that compress more than sparse1000M. Reading one byte per loop and inflating with `Z_SYNC_FLUSH` is the safest option but it is slow.

@rzinsly `nx_inflate_()` shouldn't be overflowing too much data to `fifo_in`.  That's part of the issue in #151 . I plan to fix this with PR #150.  With that said: I think it's safe for you to read as much input as possible.

File: lib/nx_gzlib.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (mscastanho) [reply lib/nx_gzlib.c:51]: > gzFile is a pointer to the semi-opaque structure gzFile_s

You're right, I missed this part. Thanks for clarifying.

File: lib/nx_gzlib.c (hunk 8 above)
Comment (rzinsly) [reply lib/nx_gzlib.c:80]: Ok, it's unlikely to happen but it can, I'll change that.

File: test/test_gz.c [hunk 14]
```diff
@@ -0,0 +1,189 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include "test.h"
+#include "test_utils.h"
+
+#define DATALEN 1024*1024 // 1MB
+#define BUFLEN	8192
+
+//const char filename[14] = "test_gzfile.gz";
```
Comment (mscastanho) [inline test/test_gz.c:10]: Delete this comment

File: test/Makefile.am [hunk 15]
```diff
@@ -86,3 +88,6 @@ regen-abi:
 	@echo "    abidw $(ABIDW_FLAGS) /path/to/libz.so > libz.abi"
 
 build-TESTS: $(check_PROGRAMS) $(check_SCRIPTS)
+
+clean-local:
+	-rm -f test_gzfile*.gz
```
Comment (mscastanho) [inline test/Makefile.am:93]: This rule is not needed since the test is already deleting the `test_gzfile*.gz` files.

File: test/test_gz.c [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+
+	return TEST_OK;
+}
+
+int test_gzwrite(Byte *src, int size, char *filename)
+{
+	gzFile file;
+	Byte *cur = src;
+	int rc;
+	char buf[BUFLEN];
```
Comment (mscastanho) [inline test/test_gz.c:73]: Why is this buffer used to read the data from `src` here? Can't we pass `src` directly to `gzwrite` like in `test_nx_gzwrite`?

**Review by mscastanho — CHANGES_REQUESTED** [review]
The test fails to compile with `-Wall -Werror`:
```
test_gz.c: In function ‘main’:
test_gz.c:143:6: error: unused variable ‘pid’ [-Werror=unused-variable]
  143 |  int pid;
      |      ^~~
test_gz.c:141:21: error: unused variable ‘compr_len’ [-Werror=unused-variable]
  141 |  const unsigned int compr_len = DATALEN*2;
      |                     ^~~~~~~~~
```

File: test/Makefile.am (hunk 15 above)
Comment (rzinsly) [reply test/Makefile.am:93]: It's just to guarantee make clean cleans everything, if the test do not finish the files will stay there.

File: test/test_gz.c (hunk 16 above)
Comment (rzinsly) [reply test/test_gz.c:73]: `test_nx_gzwrite` writes all data in one `gzwrite` call, buf if we pass `src` to `gzwrite` here each loop iteration will overwrite the data from the previous iteration.

Comment (rzinsly) [reply test/test_gz.c:73]: Anyway, we don't really need `buf` as we can pass `cur` to `gzwrite`.
I'll change that, thanks!

File: test/test_gz.c [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+
+	return TEST_OK;
+}
+
+int test_gzread(Byte *src, Byte *dest, int src_size, char *filename)
+{
+	gzFile file;
+	Byte *cur = dest;
+	int rc;
+	char buf[BUFLEN];
```
[the remark this answers is not in the document]
Comment (rzinsly) [reply test/test_gz.c]: Fixed!

File: lib/nx_gzlib.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (rzinsly) [reply lib/nx_gzlib.c:196]: I added a source code comment, `lseek` is not an option because `gzread` needs to be able to read from `stdin`.

File: lib/nx_gzlib.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (mscastanho) [reply lib/nx_gzlib.c:62]: Should check if `next_out == NULL`.

File: lib/nx_gzlib.c [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+int nx_gzread(gzFile file, void *buf, unsigned len)
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	int ret, err;
+	const uInt max = len > 10? 10 : 1;
+	uLong last_total_out = stream->total_out;
+	unsigned char* next_in = malloc(sizeof(char)*max);
```
Comment (mscastanho) [inline lib/nx_gzlib.c:169]: Should check if `next_in == NULL`

File: lib/nx_gzlib.c [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+	return stream->total_out - last_total_out;
+}
+
+int nx_gzwrite (gzFile file, const void *buf, unsigned len)
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	int rc;
+	uLong last_total_in = stream->total_in;
+	unsigned char* next_out = malloc(sizeof(char)*len);
```
Comment (mscastanho) [inline lib/nx_gzlib.c:208]: Should check if `next_out == NULL`

**Review by mscastanho — CHANGES_REQUESTED** [review]
Just some minor suggestions. After that, LGTM.

File: lib/nx_gzlib.c [hunk 20]
```diff
@@ -162,6 +162,49 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
...
+			err = nx_inflate(stream, Z_FINISH);
+			break;
+		}
+
+		stream->avail_in = ret;
+		err = nx_inflate(stream, Z_NO_FLUSH);
+
+		if (err != Z_OK && err != Z_STREAM_END){
+			state->err = err;
+			return 0;
```
Comment (tuliom) [inline lib/nx_gzlib.c:193]: Returning here will cause a memory leak for `next_in`.

**Review by tuliom — CHANGES_REQUESTED** [review]
@rzinsly I'm afraid there are still 3 leaks here. Otherwise, LGTM.

File: lib/nx_gzlib.c [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+	int ret;
+	unsigned char* next_out = malloc(sizeof(char)*MAX_LEN);
+
+	if (next_out == NULL)
+		return Z_MEM_ERROR;
+
+	if (file == NULL){
+		errno = EINVAL;
+		return Z_STREAM_ERROR;
+	}
```
[the remark this answers is not in the document]
Comment (rzinsly) [reply lib/nx_gzlib.c:70]: Done!

File: lib/nx_gzlib.c (hunk 20 above)
Comment (rzinsly) [reply lib/nx_gzlib.c:203]: Fixed by freeing it there.

File: lib/nx_gzlib.c [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+		return NULL;
+	memset(state, 0, sizeof(struct gz_state));
+
+	if (path == NULL){
+		if(fdopen(fd, mode) != NULL)
+			state->fd = fd;
+		else
+			state->fd = -1;
+	} else {
+		FILE* fp = fopen(path, mode);
```
[the remark this answers is not in the document]
Comment (rzinsly) [reply lib/nx_gzlib.c]: Done!

**Comment by rzinsly** [comment]
I rebased the code on top of current develop and was able to change the input size of `gzread` to half `len` making it way faster.
I also removed the `deflate` boolean from `gz_state` in the third commit to be able to set it when we use the sw `gzopen`.

**Review by tuliom — APPROVED** [review]
